primitives/core/src/crypto.rs

```diff
@@ -233,25 +244,46 @@ pub trait Ss58Codec: Sized + AsMut<[u8]> + AsRef<[u8]> + Default {
 				_ => Err(PublicError::UnknownVersion),
 			})
 	}
+
 	/// Some if the string is a properly encoded SS58Check address.
 	#[cfg(feature = "std")]
 	fn from_ss58check_with_version(s: &str) -> Result<(Self, Ss58AddressFormat), PublicError> {
+		const CHECKSUM_LEN: usize = 2;
 		let mut res = Self::default();
-		let len = res.as_mut().len();
-		let d = s.from_base58().map_err(|_| PublicError::BadBase58)?; // failure here would be invalid encoding.
-		if d.len() != len + 3 {
-			// Invalid length.
-			return Err(PublicError::BadLength);
-		}
-		let ver = d[0].try_into().map_err(|_: ()| PublicError::UnknownVersion)?;
 
-		if d[len + 1..len + 3] != ss58hash(&d[0..len + 1]).as_bytes()[0..2] {
+		// Must decode to our type.
+		let body_len = res.as_mut().len();
+
+		let data = s.from_base58().map_err(|_| PublicError::BadBase58)?;
+		if data.len() < 2 { return Err(PublicError::BadLength); }
+		let (prefix_len, ident) = match data[0] {
+			0..=63 => (1, data[0] as u16),
+			64..=127 => {
+				// weird bit manipulation owing to the combination of LE encoding and missing two bits
+				// from the left.
+				// d[0] d[1] are: 01aaaaaa bbcccccc
+				// they make the LE-encoded 16-bit value: aaaaaabb 00cccccc
+				// so the lower byte is formed of aaaaaabb and the higher byte is 00cccccc
```

Comment:
reading from the spec:

> The low 6 bits of this byte should be treated as the upper 6 bits of a 14 bit identifier value, with the lower 8 bits defined by the following byte.

I would have think that the upper 6 bits were the 6 strongest bits of the 14 bit identifier, thus resulting in (LE)  bbcccccc 00aaaaaa.

But implementation seems fine and is consistent

Reply:
spec should probably be changed then to keep everything in line with SCALE.

Reply:
I did run into this with my implementation since I just blindly followed SCALE 2-byte encoding instead of reading the spec carefully. (And then subsequently adjusted to match the encoding here)

I don't believe it make a huge difference, the extraction is simple, however it could be confusing from an overall consistency perspective.